REST API: Add sync related endpoint capabilities

fgiannar opened the pull request:
This PR is part of our work towards migrating endpoints to allow blog tokens described in #17379 and covers the two sub-tasks under the Technical Debt section.
In more detail, in order to allow blog token usage in the past we had removed the Sync related endpoint capabilities in #16893 
With #17331 though, we can now both have needed capabilities defined for endpoints (if a user token is used) AND allow blog token usage, as long as the endpoint has the `allow_jetpack_site_auth` flag set.
So in this PR, we are adding again the needed capabilities we had removed from Sync related endpoints.

#### Changes proposed in this Pull Request:
* Sets the `allow_jetpack_site_auth` flag for Sync related endpoints (covers Options related endpoints as well as they are used by Sync).
* Re-adds the previously removed needed permissions in Sync related endpoints (covers Options related endpoints as well as they are used by Sync).

#### Jetpack product discussion


#### Does this pull request change what data or activity we track or use?
No

#### Testing instructions:
Go to the [developer console](https://developer.wordpress.com/docs/api/console/) and perform the following requests as a site admin (checkout `json-api-jetpack-endpoints.php` for details on request parameters and sample requests):

1. `POST /sites/%s/sync`
2. `GET /sites/%s/sync/status`
3. `GET /sites/%s/sync/settings`
4. `POST /sites/%s/sync/settings`
5. `GET /sites/%s/sync/object`
6. `POST /sites/%s/sync/now`
7. `POST /sites/%s/sync/checkout`
8. `POST /sites/%s/sync/close`
9. `POST /sites/%s/sync/unlock`
10. `GET /sites/%s/data-checksums`
11. `GET /sites/%s/data-histogram`
12. `GET /sites/%s/sync/object-id-range`
13. `GET sites/%s/option?option_name=blogname`
14. `POST sites/%s/option`

- [x] Make sure you receive `20*` responses for endpoints 1-14:

* Repeat the above requests as a non-admin user of your site and confirm you get:

- [x] For endpoints 1-9: `403 Unauthorized` responses with message: "This endpoint needs more permissions from the user.".
- [x] For endpoints 10-14: `403 Unauthorized` responses with message: "This user is not authorized to manage_options on this blog.".

* Repeat the above requests as a non-user of your site and confirm you get:
- [x] For endpoints 1-9: `403 Unauthorized` responses with message: "This endpoint needs more permissions from the user.".
- [x] For endpoints 10-14: `400` responses with message: "Unknown Token.".

* Repeat the above requests without any authorization (outside of the console, eg using Postman) and confirm you get:
- [x] For endpoints 1-14: 
    - GET requests should return `403 Unauthorized` responses with message: "This endpoint needs more permissions from the user."
    - POST requests should return `405 Unauthorized` responses with message: "That method is not allowed."

* Repeat the above requests using a blog token and confirm you get:
- [x] For endpoints 1-14: `20*` responses.

#### Proposed changelog entry for your changes:
<!-- Please do not leave this empty. If no changelog entry needed, state as such. -->
* N/A :: non functional change, hardening of Jetpack Connection and Sync.

leogermani reviewed (commented):
Finally managed to test all requests! Here is what I got:

Just two small corrections
13. you must query for a allowed listed option, such as `blogname`
14. The endpoint is `option` in the singular form.

== For admin and Blog token ==

Everything worked as expected.

I couldn`t really test 7, 8 and 9 because it gave me errors such as the queue is empty. But I guess this is fine

== For non logged users ==

Everything worked as expected. Just a note that apparently the request are not even sent to the remote site.

== For a non-linked logged in user ==

Everything worked except that all `POST` request returned `405` instead of `403`.

fgiannar commented:
Hey @leogermani ,

Thank you so much for taking the time to test all aforementioned requests!

> Just two small corrections
> 13. you must query for a allowed listed option, such as blogname
> 14. The endpoint is option in the singular form.

Yes, you are right. I've updated the test instructions for consistency

> Everything worked except that all POST request returned 405 instead of 403.

Right again! I've updated the test instructions for consistency

leogermani reviewed (approved):
Good!
